### ISSUE TITLE
Add configuration option to put everything on the module path when working with Java modules

### ISSUE THREAD

**Issue by jjohannes** [gradle/gradle#12630]
Use cases for this are well described here: https://github.com/gradle/gradle/issues/890#issuecomment-603868530

```
modularClasspathHandling.<new-opition>.set(true)
```

Name suggestions:
- treatLibrariesAsAutomaticModules
- allModulePath
- noUnnamedModule

**Comment by jjohannes** [comment]
If you decide to put everything on the `--module-path` you can run into trouble with libraries that are not valid modules (invalid name, split packages - see also https://github.com/gradle/gradle/issues/890#issuecomment-604199850 point 1).

We should consider adding a basic exclusion mechanism for such libraries:

```
modularClasspathHandling.excludeAsModule('native.platform') // 'native.platform' is the, possibly invalid, module name
```

We could also consider to automatically detect invalid modules, which would not be too hard for invalid names (related #7752), but tricky for package split situations. And there might also be other reasons why one would like to exclude a certain library. So an exclude mechanism feels like the right first step.

**Comment by jjohannes** [comment]
We had some discussion about this topic in the team and we are currently not sure if or what options we really need to add here.

I [added a sample](https://docs.gradle.org/release-nightly/samples/sample_java_modules_with_transform.html) that shows how Gradle's artifact transforms can be used to enrich Jars without module information. I think for some of the use cases that were described this is a nice solution because it allows you to even deal with modules that have an invalid auto-name. Or go even further to solve a split-package issue.

Anyone interested in this topic, please have a look at the solution provided in the sample.

If you think this is not sufficient, it would be very helpful if you could describe a full use case, ideally including real (problematic) libraries out there. Then we can discuss this further.

**Comment by iherasymenko** [comment]
@jjohannes thank you for this. I couldn't ask for more :-)

**Comment by JojOatXGME** [comment]
@jjohannes Using artifact transforms looks like a good solution for incompatible libraries. Beside that, I was also concerned about plugins that have their own handling for the module path. This might include existing plugins which get updated to Gradle 6.4. I think for such plugins, a `modulePath` property on the relevant tasks would be all they need. This property would then be empty by default.

However, I'm not sure if such option would be necessary. Maybe JPMS plugins are fine with just using the behavior of Gradle. However, I'm a little bit worried that the automatic detection of Gradle might become problematic when combined with other options like `--patch-module`.

**Comment by TheMrMilchmann** [comment]
Another use-case for putting non-modular dependencies on the module path is the ability to work easily with non-modular/old compile-time only dependencies. Specifically, I still use jsr305 annotations extensively. However, since the artifact is quite old, it doesn't even have an "Automatic-Module-Name" entry in its manifest and `inferModulePath` does not recognize it as a module. Since named modules cannot read the unnamed module, the library remains inaccessible. Instead, it should be possible to write a module descriptor as follows:

```java
module com.example.module {
    requires static jsr305;
}
```

Currently, I use the [Java 9 Modularity plugin](https://github.com/java9-modularity/gradle-modules-plugin) which solves this by placing all dependencies on the module path. (However, this does not cover all use-cases either and I would rather be able to override whether a library is placed on the module path or the classpath on a case-by-case basis. See also https://github.com/java9-modularity/gradle-modules-plugin/issues/90.)

(While it is not a good idea to declare dependencies on Automatic-Modules with inferred names, it is perfectly safe for compile-time only dependencies since they are neither exposed to dependant modules nor required to build them.)


On another note, I have been using `requires static` to work with non-modularized libraries in combination with manually adjusting the command-line args of related tasks to make them visible for a desktop application which makes use of jlink for the vast majority of its dependencies. (Again, this is safe since the application is not meant to be consumed by others.)

While the artifact transform sample linked above does seem to do the trick in both cases (though I haven't tested the latter extensively yet), I would love to have a "simpler" way to configure where those libraries should be placed without messing with the artifacts themselves.

**Comment by jjohannes** [comment]
More feedback regarding the use of legacy libraries as modules: https://github.com/gradle/gradle/issues/890#issuecomment-616919397

**Comment by msgilligan** [comment]
@jjohannes As I've said over in Issue #890, I think putting the plugin from the samples in a Github repo and releasing it on the portals plugin would be a reasonable next step and would certainly allow people to kick the tires Gradle 6.4 on some real-world projects.

**Comment by jjohannes** [comment]
@msgilligan I thought about what you wrote and the sample you gave. It's very helpful to see this in action on a real project. I decided to just publish the plugin we have now. Then I can continuously update that based on feedback independent on Gradle releases:

https://github.com/jjohannes/extra-java-module-info#how-to-use-this-plugin

In the future, we can then decide where we move that based on the use cases that emerged.

Everyone, feel free to try the plugin and give feedback.

**Comment by msgilligan** [comment]
@jjohannes As I [mentioned](https://github.com/gradle/gradle/issues/890#issuecomment-617999414) in Issue #890, my sample (a branch of [ConsensusJ](https://github.com/ConsensusJ/consensusj)) is now using the published plugin and working well! [The diff](https://github.com/ConsensusJ/consensusj/commit/ae1bd6a993658f838b606b917ee3ace2eefb83ab) from `master` is interesting as it shows the changes required to convert an automatic module to a real module. (The diff includes the upgrade to Gradle 6.4-rc-2, but will get cleaner when 6.4 is final and `master` is updated to 6.4.)

**Comment by wind57** [comment]
@jjohannes we tried the plugin and so far all of our cases were covered. We love it inside the team. Thank you for this effort!

**Comment by shutyaev** [comment]
While the plugin made by @jjohannes is pretty good it does require extra bit of configuration for each jar. At the same time the ModuleFinder.of javadoc [1] clearly indicates that a jar even without an Automatic-Module-Name manifest entry will still be treated as an automatic module with its name being derived from the jar name. As this behavior exists in jvm it would be nice to have it replicated in gradle, maybe not as the default one, but at least with an option to turn it on via a switch...

**Comment by dweiss** [comment]
I agree with @shutyaev - the automatic module detector currently present in gradle is not always sufficient. Sometimes you'd want the Java layer to handle the conversion - it's why it's there in the first place. So some way of telling "hey, this and that jar/ dependency should be placed on module path, even if it doesn't look like a module" would be really useful.
